pkg/quota/evaluator/core/persistent_volume_claims.go

```diff
@@ -51,68 +54,127 @@ func listPersistentVolumeClaimsByNamespaceFuncUsingClient(kubeClient clientset.I
-		UsageFunc:            PersistentVolumeClaimUsageFunc,
-		ListFuncByNamespace:  listFuncByNamespace,
+		MatchesScopeFunc:    generic.MatchesNoScopeFunc,
+		ConstraintsFunc:     constraintsFunc,
+		UsageFunc:           usageFunc,
+		ListFuncByNamespace: listFuncByNamespace,
 	}
 }
 
-// PersistentVolumeClaimUsageFunc knows how to measure usage associated with persistent volume claims
-func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
-	result := api.ResourceList{}
-	var found bool
-	var request resource.Quantity
-
-	switch t := object.(type) {
+func toInternalPersistentVolumeClaimOrDie(obj runtime.Object) *api.PersistentVolumeClaim {
+	pvc := &api.PersistentVolumeClaim{}
+	switch t := obj.(type) {
 	case *v1.PersistentVolumeClaim:
-		request, found = t.Spec.Resources.Requests[v1.ResourceStorage]
+		if err := v1.Convert_v1_PersistentVolumeClaim_To_api_PersistentVolumeClaim(t, pvc, nil); err != nil {
+			panic(err)
+		}
 	case *api.PersistentVolumeClaim:
-		request, found = t.Spec.Resources.Requests[api.ResourceStorage]
+		pvc = t
 	default:
-		panic(fmt.Sprintf("expect *api.PersistenVolumeClaim or *v1.PersistentVolumeClaim, got %v", t))
+		panic(fmt.Sprintf("expect *api.PersistentVolumeClaim or *v1.PersistentVolumeClaim, got %v", t))
 	}
+	return pvc
+}
 
+// PersistentVolumeClaimUsageFunc knows how to measure usage associated with persistent volume claims
+func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
+	result := api.ResourceList{}
+	pvc := toInternalPersistentVolumeClaimOrDie(object)
+	storageClassRef := util.GetClaimStorageClass(pvc)
+
+	// charge for claim
 	result[api.ResourcePersistentVolumeClaims] = resource.MustParse("1")
-	if found {
+	if len(storageClassRef) > 0 {
+		storageClassClaim := api.ResourceName(storageClassRef + storageClassSuffix + string(api.ResourcePersistentVolumeClaims))
+		result[storageClassClaim] = resource.MustParse("1")
+	}
+
+	// charge for storage
+	if request, found := pvc.Spec.Resources.Requests[api.ResourceStorage]; found {
 		result[api.ResourceRequestsStorage] = request
+		// charge usage to the storage class (if present)
+		if len(storageClassRef) > 0 {
+			storageClassStorage := api.ResourceName(storageClassRef + storageClassSuffix + string(api.ResourceRequestsStorage))
+			result[storageClassStorage] = request
+		}
 	}
+
 	return result
 }
 
-// PersistentVolumeClaimConstraintsFunc verifies that all required resources are present on the claim
-// In addition, it validates that the resources are valid (i.e. requests < limits)
-func PersistentVolumeClaimConstraintsFunc(required []api.ResourceName, object runtime.Object) error {
-	pvc, ok := object.(*api.PersistentVolumeClaim)
-	if !ok {
-		return fmt.Errorf("unexpected input object %v", object)
-	}
+// makePersistentVolumeClaimConstraintsFunc returns a function that knows how to enforce constraints.
+func makePersistentVolumeClaimConstraintsFunc(usageFunc generic.UsageFunc) generic.ConstraintsFunc {
```

Comment:
I think a function factory function which accepts another function to be "generic" is a bit too far.  This really is a special snowflake.